mlflow/utils/search_utils.py

```diff
@@ -1,45 +1,198 @@
-        elif metric_type == 'double':
-            comparator = search_expression.metric.double.comparator
-            value = search_expression.metric.double.value
+import sqlparse
+from sqlparse.sql import Identifier, Token, Comparison, Statement
+from sqlparse.tokens import Token as TokenType
+
+from mlflow.exceptions import MlflowException
+from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
+
+
+class SearchFilter(object):
+    VALID_METRIC_COMPARATORS = set(['>', '>=', '!=', '=', '<', '<='])
+    VALID_PARAM_COMPARATORS = set(['!=', '='])
+    _METRIC_IDENTIFIER = "metric"
+    _ALTERNATE_METRIC_IDENTIFIERS = set(["metrics"])
+    _PARAM_IDENTIFIER = "parameter"
+    _ALTERNATE_PARAM_IDENTIFIERS = set(["param", "params"])
+    VALID_KEY_TYPE = set([_METRIC_IDENTIFIER] + list(_ALTERNATE_METRIC_IDENTIFIERS)
+                         + [_PARAM_IDENTIFIER] + list(_ALTERNATE_PARAM_IDENTIFIERS))
+    VALUE_TYPES = set([TokenType.Literal.String.Single,
+                       TokenType.Literal.Number.Integer,
+                       TokenType.Literal.Number.Float])
+
+    def __init__(self, search_runs=None):
+        self._filter_string = search_runs.filter if search_runs else None
+        self._search_expressions = search_runs.anded_expressions if search_runs else None
+        if self._filter_string and self._search_expressions:
+            raise MlflowException("Can specify only one of 'filter' or 'search_expression'",
+                                  error_code=INVALID_PARAMETER_VALUE)
+
+        # lazy parsing
+        self.parsed = None
+
+    @property
+    def filter_string(self):
+        return self._filter_string
+
+    @property
+    def search_expressions(self):
+        return self._search_expressions
+
+    @classmethod
+    def _trim_backticks(cls, entity_type):
+        if entity_type.startswith("`"):
+            assert entity_type.endswith("`")
+            return entity_type[1:-1]
+        return entity_type
+
+    @classmethod
+    def _valid_entity_type(cls, entity_type):
+        entity_type = cls._trim_backticks(entity_type)
+        if entity_type not in cls.VALID_KEY_TYPE:
+            raise MlflowException("Invalid search expression type '%s'. "
+                                  "Valid values are '%s" % (entity_type, cls.VALID_KEY_TYPE))
+
+        if entity_type in cls._ALTERNATE_PARAM_IDENTIFIERS:
+            return cls._PARAM_IDENTIFIER
+        elif entity_type in cls._ALTERNATE_METRIC_IDENTIFIERS:
+            return cls._METRIC_IDENTIFIER
         else:
-            raise Exception("Invalid metric type: '%s', expected float or double")
-        metric = next((m for m in run.data.metrics if m.key == key), None)
-        if metric is None:
+            # either "metric" or "parameter", since valid type
+            return entity_type
+
+    @classmethod
+    def _get_identifier(cls, identifier):
+        try:
+            entity_type, key = identifier.split(".")
```

Comment:
It looks like this throws a ValueError if there's more than one "." in the search filter because the result of split() will contain more than two elements (so e.g. use cases like searching on a param with a period in the value, like `metrics.modelClassName = 'sklearn.ensemble.RandomForestClassifier'`, will break). Is that intentional & if so should we document it in the protobuf so it shows up in public docs?

Reply:
No. This is applicable only for the left hand side if the expression. So if the filter were defined like

```
metric.crazy.name <= 0.01
```

And will error out. Will clean up message and add a test case.